### PR TITLE
Serialize all semesters enrolled

### LINKED ISSUE #3840: Misleading semester enrolled search results
Related to #3810 

The way enrollments are serialized here only the first found paid course run enrollment is serialized and indexed, and that is the course run grade and semester that will show up in the search results.

### PR DESCRIPTION
**Pull request by annagav**
#### What are the relevant tickets?
Fix #3840

#### What's this PR do?
Serialize  all semester enrollments for users that took the same course multiple times in different semesters.

#### How should this be manually tested?
Have your user enrolled in the same course multiple times, with paid/audited variation. When the user comes up in the search all semesters that the user was enrolled in show up in the Semester facet.

### CODE REVIEW

File: dashboard/serializers.py [hunk 1]
```diff
@@ -49,14 +49,27 @@ def serialize_enrollment(cls, mmtrack, course_run):
...
-        semester = cls.serialize_semester(course_run)
         return {
             'final_grade': final_grade.grade_percent if final_grade else None,
-            'semester': semester,
             'course_title': course_title,
             'payment_status': payment_status,
         }
 
+    @classmethod
+    def semesters_enrolled(cls, mmtrack):
```
Comment (gsidebo) [inline dashboard/serializers.py]: Should be called `serialize_semesters_enrolled` or `serialize_enrolled_semesters`

**Comment by annagav** [comment]
@gsidebo I think I addressed all comments

**Comment by gsidebo** [comment]
@annagav @pdpinch I was just testing this change and wanted to make sure that this behavior is what we want. My concern is that the 'Course' facet and the 'Semester' facet are basically independent from each other now. If I selected the 'Analog Learning 100' course and the '2015 - Summer' semester, I would expect my search results to show me learners that were enrolled in the Summer 2015 course run of 'Analog Learning 100'. What it shows instead are users that have been enrolled in any 'Analog Learning 100' at some point, and who also have been enrolled in **any** Summer 2015 course (not necessarily the 'Analog Learning 100' run). 

ss 2018-04-30 at 16 15 13

ss 2018-04-30 at 16 37 02

File: dashboard/serializers.py [hunk 2]
```diff
@@ -49,14 +49,27 @@ def serialize_enrollment(cls, mmtrack, course_run):
...
+
+        Args:
+            mmtrack (MMTrack): An MMTrack object
+        Returns:
+            list: Serialized all semester enrollments
+        """
+        serialized_enrollments = []
+        for course_run in mmtrack.get_all_enrolled_course_runs():
+            serialized_enrollments.append({'semester': cls.serialize_semester(course_run)})
+        return serialized_enrollments
```
[the remark this answers is not in the document]
Comment (gsidebo) [reply dashboard/serializers.py]: We can still change this to a dict comprehension

**Comment by pdpinch** [comment]
Can we add a selenium test for this?

**Comment by gsidebo** [comment]
@pdpinch Here's an example scenario:

I have an enrollment in the Fall 2017 run of Course A and the Spring 2018 run of Course B. If an instructor selects the 'Course A' course filter, and the 'Spring 2018' semester filter, I will show up in the search results **even though I was never enrolled in the Course A Spring 2018 course run**. Those filters only take into account that I have (a) an enrollment in any course run for 'Course A', and (b) an enrollment in any Spring 2018 course run, regardless of the course.

**Comment by pdpinch** [comment]
So it doesn't expand the result set, but it also doesn't narrow the result set as much as you might expect. 

Do you have any thoughts on how me might make this behave differently? Or, alterately, make it clearer what the behavior is?

**Comment by gsidebo** [comment]
@pdpinch My first thought was that Course/Semester should be a hierarchical menu a la Current Residence (which filters by state/region). I can't say for sure how big a lift that would be in terms of the index structure or how performance would be affected. I think this just come downs to our users' expectations. The behavior I described above doesn't match what I would expect, but I don't know how our users see it. 

Here's a good question to ask at this point: would our users get any value from using this semester filter on its own with no other filters? Example: would any of our instructors have  a need to search for all students that had _any_ enrollment in _any_ course in Fall 2017? If the answer is 'no', the behavior should probably change. If the answer is 'yes', then we should probably keep this as-is, and clear up the messaging. The `Semester` heading could be changed to something like `Enrolled Semester`/`Active Semester`/`Semester with Enrollment`/etc

**Comment by pdpinch** [comment]
If you look at #3913 you'll see the use case we had in mind: the course team wants to send an email to everyone who has enrolled in a (DEDP) course in 2017. 

Given that use case, I understand your recommendation to be that we should change the filter label to something like "Enrolled Semester". That makes sense to me.

**Comment by pdpinch** [comment]
What do you think about disabling the semester facet when he course facet is in use; and vice versa?

**Comment by gsidebo** [comment]
@pdpinch @annagav That would work. I'm not sure if we're doing that with any other facets though, and I don't remember Searchkit supporting that kind of feature out of the box. Would certainly add some development time. Would it be better to put the feature out there and see if it's too confusing to our users?

**Comment by pdpinch** [comment]
@gsidebo you labelled this as "waiting on author" but you still haven't approved your review yet. Is there something more @annagav needs to do here?

**Comment by annagav** [comment]
@gsidebo I renamed the facet.

**Comment by gsidebo** [comment]
@annagav Selected filter label is unchanged:

ss 2018-05-09 at 15 33 42

Also FWIW I'm not sure about "Enrolled Semester". I think an instructor who has applied the filters "Course - Analog Learning 100" and "Enrolled Semester - Spring 2016" is still going to think they are only looking at students who were enrolled in the Spring 2016 run of Analog Learning 100. I can't think of anything more helpful than "Active Semester" or less verbose than "Semester with Course Enrollments", so maybe it's best just to deploy this change and see if anyone complains (cc @pdpinch)

**Pull request merged**